src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -937,26 +940,63 @@
                                                      "code_id"
                                                      (current-code-id-fn (get request-options "environment")))))})))
 
+(defn parse-project-compile-data
+  "Parse data required to compile a catalog inside a project. Data required includes
+   * Root path to project
+   * modulepath
+   * hiera config
+   * project_name"
+  [request-options
+   bolt-projects-dir]
+  (let [versioned-project (:versioned_project request-options)
+        project-root (file-serving/get-project-root bolt-projects-dir versioned-project)
+        project-config (file-serving/read-bolt-project-config project-root)]
+    (-> request-options
+        (assoc "project_root" project-root)
+        (assoc "modulepath" (file-serving/get-project-modulepath project-config))
+        (assoc "hiera_config" (:hiera-config project-config))
+        (assoc "project_name" (:name project-config)))))
+
 (schema/defn ^:always-validate compile-fn :- IFn
   [jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)
    current-code-id-fn :- IFn
-   boltlib-path :- (schema/maybe [schema/Str])]
+   boltlib-path :- (schema/maybe [schema/Str])
+   bolt-projects-dir :- (schema/maybe schema/Str)]
   (fn [request]
-    (let [env (jruby-request/get-environment-from-request request)
-          request-options (-> request
+    (let [request-options (-> request
                               :body
                               slurp
                               (validated-body CompileRequest))
-          compile-options (assoc request-options
-                                 "code_id"
-                                 (current-code-id-fn env))]
-      {:status 200
-       :headers {"Content-Type" "application/json"}
-       :body (json/encode
-               (jruby-protocol/compile-ast jruby-service
-                                           (:jruby-instance request)
-                                           compile-options
-                                           boltlib-path))})))
+          versioned-project (get request-options :versioned_project)
+          environment (get request-options :environment)]
```

Comment:
I'm not convinced this is how you would get environment... I have not dug in yet, but it appears this is always null based on EVERY compile request (from environments) failng with this message.